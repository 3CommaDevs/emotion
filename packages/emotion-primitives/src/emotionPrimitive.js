import React from 'react'
import primitives from 'react-primitives'
import { renderStylesOnNative } from './emotion-native'

const Platform = primitives.Platform

const TAGS = ['Text', 'View', 'Image']

/*
  Emotion primitive - Use this function instead of 'styled' to render the styles using emotion on Web, React Native and Sketch.

  Tradeoffs -

  * Passing props in string template literal is not supported yet when used with React Native. But you can pass props when using the object styles.
  * Compute time is more so the render on Sketch is slow right now.
  * Need to tweak the webpack config when used with Sketch in order suppress warnings for 'fs' module.
  * Cannot use shorthand because it requires babel plugin.
  * Need to pass primitives[tag_name] when withComponent is used.

  Usage -

  const Text = emotionPrimitive('Text')`
    color: hotpink;
    font-size: 30px;
  `

  render(<Text>Hello</Text>)

  const View = emotionPrimitive('View')({
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  }, (props) => {
    color: props.color
  })

  render(<View color='mistyrose' />)

  const Title = Text.withComponent('Text')

  render(<Title>Hello</Title>)
*/
export const emotionPrimitive = tag => {
  if (Platform.OS === 'web') {
    const styled = require('react-emotion').default

    if (typeof tag !== 'string') {
      throw new Error('Expected primitive name to be a string')
    }

    if (TAGS.includes(tag)) {
      // TODO: Assign primitives directly to styled constructor
      return styled(primitives[tag])
    }

    throw new Error(
      `${tag} is an invalid primitive. Expected primitive to be one of ['Text', 'View', 'Image'].`
    )
  } else {
    // For React Native and Sketch
    return (...args) => renderStylesOnNative(tag, args)
  }
}
