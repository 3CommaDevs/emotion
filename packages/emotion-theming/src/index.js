// @flow
export { createThemeProvider } from './theme-provider'
export { createWithTheme } from './with-theme'
export { channel, contextTypes } from './utils'
