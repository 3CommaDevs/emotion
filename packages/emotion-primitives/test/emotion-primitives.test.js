import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { ThemeProvider, withTheme } from 'emotion-theming'

import emotionPrimitive from '../src'

Enzyme.configure({ adapter: new Adapter() })

const theme = { backgroundColor: 'magenta' }

describe('Emotion primitives', () => {
  test('should not throw an error when used valid primitive', () => {
    expect(() => emotionPrimitive.text({})).not.toThrow()
  })

  test('should throw an error when used invalid primitive', () => {
    expect(() => emotionPrimitive.TEXT({})).toThrowError(
      `Cannot style invalid primitive TEXT. Expected primitive to be one of ['Text', 'View', 'Image']`
    )
  })

  test('theme', () => {
    const T = withTheme(emotionPrimitive.text`color: red; backgroundColor: ${props => props.theme.backgroundColor}`)

    const tree = renderer.create(<ThemeProvider theme={theme}><T bg='red'>Hello</T></ThemeProvider>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('should render the primitive when styles applied using object style notation', () => {
    const Text = emotionPrimitive.text`color: red; font-size: 20px; background-color: ${props =>
      props.back};`
    const tree = renderer
      .create(
        <Text back="red" fontSize={40}>
          Emotion Primitives
        </Text>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should render the primitive on changing the props', () => {
    const Text = emotionPrimitive.text({ padding: '20px' }, props => ({
      color: props.decor
    }))
    const tree = renderer
      .create(<Text decor="hotpink">Emotion Primitives</Text>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should render primitive with style prop', () => {
    const Title = emotionPrimitive.text`color: hotpink;`
    const tree = renderer
      .create(<Title style={{ padding: 10 }}>Emotion primitives</Title>)
      .toJSON()
    expect(tree).toMatchSnapshot()
    const wrapper = Enzyme.shallow(
      <Title style={{ padding: 10 }}>Emotion primitives</Title>
    )
    expect(wrapper.find('Text').prop('style')).toEqual([77, { padding: 10 }])
  })

  test('primitive should work with `withComponent`', () => {
    const Text = emotionPrimitive.text`color: ${props => props.decor};`
    const Name = Text.withComponent('Text')
    const tree = renderer.create(<Name decor="hotpink">Mike</Name>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should render primitive with css overrides', () => {
    const Text = emotionPrimitive.text`color: hotpink;`
    const tree = renderer.create(<Text fontSize={40}>Emotions</Text>).toJSON()
    expect(tree).toMatchSnapshot()
    const wrapper = Enzyme.shallow(<Text fontSize={20}>Emotions</Text>)
    expect(wrapper.find('Text').prop('style')).toEqual([80, { fontSize: 20 }])
  })

  it('should render <Image />', () => {
    const Image = emotionPrimitive.image`
      border: 2px solid hotpink;
    `
    const tree = renderer
      .create(
        <Image
          source={{
            uri:
              'https://camo.githubusercontent.com/209bdea972b9b6ef90220c59ecbe66d35ffefa8a/68747470733a2f2f63646e2e7261776769742e636f6d2f746b6834342f656d6f74696f6e2f6d61737465722f656d6f74696f6e2e706e67',
            height: 150,
            width: 150
          }}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
