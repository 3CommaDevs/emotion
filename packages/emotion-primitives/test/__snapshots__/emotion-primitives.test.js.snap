// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emotion primitives primitive should work with \`withComponent\` 1`] = `
<div
  className="emotion-0 rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb emotion-1 rn-boxSizing-deolkf rn-color-k9pu7p rn-display-1471scf rn-font-1lw9tu2 rn-fontFamily-10u92zi emotion-2 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi emotion-3 rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j emotion-4 emotion-5 emotion-6 rn-wordWrap-qvutc0"
  dir="auto"
>
  Mike
</div>
`;

exports[`Emotion primitives should render <Image /> 1`] = `
<div
  className="emotion-0 rn-backgroundColor-wib322 rn-backgroundPosition-vvn4in emotion-1 emotion-2 emotion-3 emotion-4 emotion-5 emotion-6 emotion-7 emotion-8 emotion-9 emotion-10 emotion-11 emotion-12 emotion-13 emotion-14 rn-boxSizing-deolkf rn-display-6koalj emotion-15 emotion-16 emotion-17 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi emotion-18 emotion-19 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j emotion-20 rn-position-bnwqim emotion-21"
/>
`;

exports[`Emotion primitives should render primitive with style prop 1`] = `
<div
  className="emotion-0 rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb emotion-1 rn-boxSizing-deolkf rn-color-k9pu7p rn-display-1471scf rn-font-1lw9tu2 rn-fontFamily-10u92zi emotion-2 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi emotion-3 emotion-4 emotion-5 rn-wordWrap-qvutc0"
  dir="auto"
  style={
    Object {
      "paddingBottom": "10px",
      "paddingLeft": "10px",
      "paddingRight": "10px",
      "paddingTop": "10px",
    }
  }
>
  Emotion primitives
</div>
`;

exports[`Emotion primitives should render the primitive on changing the props 1`] = `
<div
  className="emotion-0 rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb emotion-1 rn-boxSizing-deolkf rn-display-1471scf rn-font-1lw9tu2 rn-fontFamily-10u92zi emotion-2 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi emotion-3 rn-paddingTop-1knelpx rn-paddingRight-1ah4tor rn-paddingBottom-k8qxaj emotion-4 emotion-5 emotion-6 rn-wordWrap-qvutc0"
  dir="auto"
  style={
    Object {
      "color": "hotpink",
    }
  }
>
  Emotion Primitives
</div>
`;

exports[`Emotion primitives should render the primitive when styles applied using object style notation 1`] = `
<div
  className="rn-backgroundColor-1mjtqww emotion-0 rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb emotion-1 rn-boxSizing-deolkf emotion-2 rn-display-1471scf rn-font-1lw9tu2 rn-fontFamily-10u92zi rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi emotion-3 rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j emotion-4 emotion-5 emotion-6 rn-wordWrap-qvutc0"
  dir="auto"
  style={
    Object {
      "fontSize": "40px",
    }
  }
>
  Emotion Primitives
</div>
`;

exports[`Emotion primitives should style any other component 1`] = `
<div
  className="emotion-0 rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb emotion-1 rn-boxSizing-deolkf rn-color-k9pu7p rn-display-1471scf rn-font-1lw9tu2 rn-fontFamily-10u92zi emotion-2 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi emotion-3 rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j emotion-4 emotion-5 emotion-6 rn-wordWrap-qvutc0"
  dir="auto"
>
  Hello World
</div>
`;

exports[`Emotion primitives should work with emotion-theming 1`] = `
<div
  className="emotion-0 rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb emotion-1 rn-boxSizing-deolkf rn-color-ar5qwb rn-display-1471scf rn-font-1lw9tu2 rn-fontFamily-10u92zi emotion-2 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi emotion-3 rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j emotion-4 emotion-5 emotion-6 rn-wordWrap-qvutc0"
  dir="auto"
>
  Hello World
</div>
`;
